lib/logic/player/content.dart

```diff
@@ -23,137 +22,71 @@ extension QuickActionSerialization on QuickAction {
-///
-/// The point of this function is to structurize and generalize the places where multiple contents
-/// an be used. It also allows to ensure that every existing content in the app is supported in all
-/// places it should be supported, this is extra useful when new content type is added.
-///
-/// Sometimes this function cannot be used directly to handle the content-related logic.
-/// Then we put a copy-paste assertion like below, so we don't forget to update the
-/// related code around this assertion, when we add a new content type.
-///
-/// An example of this can be a [ContentTuple].
-///
-/// ```
-/// assert(() {
-///   // See contentPick documentation for why we need this.
-///   contentPick<Song, void>(
-///     song: null,
-///     album: null,
-///     playlist: null,
-///     artist: null,
-///   );
-///   return true;
-/// }());
-/// ```
-///
-V contentPick<T extends Content, V>({
-  Type? contentType,
-  required V song,
-  required V album,
-  required V playlist,
-  required V artist,
-  V? fallback,
-}) {
-  switch (contentType ?? T) {
-    case Song:
-      return song;
-    case Album:
-      return album;
-    case Playlist:
-      return playlist;
-    case Artist:
-      return artist;
-    case Content:
-      if (fallback != null) {
-        return fallback;
-      }
-      throw UnimplementedError();
-    default:
-      throw UnimplementedError();
-  }
-}
-
-/// Analogue of [contentPick] for [PersistentQueue]s.
-V persistentQueuePick<T extends PersistentQueue, V>({
-  Type? contentType,
-  required V album,
-  required V playlist,
-  V? fallback,
-}) {
-  switch (contentType ?? T) {
-    case Album:
-      return album;
-    case Playlist:
-      return playlist;
-    case Content:
-      if (fallback != null) {
-        return fallback;
-      }
-      throw UnimplementedError();
-    default:
-      throw UnimplementedError();
-  }
-}
-
-/// A [Map] container for the [Content] as key, and `V` as value entry.
+/// A [Map] container for the [ContentType] as key, and [V] as value entry.
 class ContentMap<V> {
-  /// Creates a content map from initial value [map].
-  ///
-  /// If none specified, will initialize with empty map.
-  ContentMap([Map<Type, V>? map]) : _map = map ?? {};
-
-  final Map<Type, V> _map;
-
-  /// Map values.
-  Iterable<V> get values => _map.values;
-
-  /// Map entries.
-  Iterable<MapEntry<Type, V>> get entries => _map.entries;
-
-  /// Returns a value per `T` [Content] from the map.
-  ///
-  /// If [key] was explicitly provided, will use it instead.
-  V? getValue<T extends Content>([Type? key]) {
-    assert(
-      Content.enumerate().contains(key ?? T),
-      "Specified type must be a subtype of Content",
+  /// The value for [ContentType.song].
+  V songValue;
+  /// The value for [ContentType.album].
+  V albumValue;
+  /// The value for [ContentType.playlist].
+  V playlistValue;
+  /// The value for [ContentType.artist].
+  V artistValue;
```

Comment:
This should be a `Map<ContenType, V>`. This would greatly reduce the implementation size.

Reply:
I would like to avoid using a `Map` here and in the `ContentTuple`, because

* A `Map` is (very slightly, but still) less efficient, accessing an item requires a runtime lookup. With the current way the compiler can resolve the item at compile time.
* Using a `Map` we would loose some of our type guarantees. We know that there is exactly one item for each content type, and the compiler should know too. With a map, we loose that information and have to force unwrap the value on access (e.g. `map[contentType]!`)
* At least for the `ContentTuple`, we still want the `songs`, `albums` ... accessors, so the code is actually not going to be much smaller.
* The explicit specification of a field for each content type makes sure we don'T forget to add a constructor argument when we add a new `ContentType`. Right now, the compiler would complain on the `switch` in `get` and `set`, which forces us to add a new field, which needs to be initialized so it forces us to add a new constructor argument. With a map, this chain would end at the `switch`.